src/storage/txn/scheduler.rs

```diff
@@ -845,25 +847,39 @@ impl<E: Engine, L: LockManager> Scheduler<E, L> {
                                 KV_COMMAND_KEYWRITE_HISTOGRAM_VEC
                                     .get(tag)
                                     .observe(rows as f64);
+
+                                if ok {
+                                    // consume the quota only when write succeeds.
+                                    if sched.inner.flow_controller.enabled() {
+                                        if sched.inner.flow_controller.is_unlimited() {
+                                            // no need to delay if unthrottled, just call consume to record write flow
+                                            let _ = sched.inner.flow_controller.consume(write_size);
+                                        } else {
+                                            // Control mutex is used to ensure there is only one request consuming the budget.
+                                            // The delay may exceed 1s, and the speed limit is changed every second.
+                                            // If the speed of next second is larger than the one of first second,
+                                            // without the mutex, the write flow can't throttled strictly.
+                                            let _guard = sched.control_mutex.lock().await;
+                                            let delay =
+                                                sched.inner.flow_controller.consume(write_size);
```

Comment:
comment why we need to consume budget in the finish callback, and relay the delay to the other writes via the mutex lock.